pyccel/codegen/printing/ccudacode.py

```diff
@@ -0,0 +1,591 @@
+            else:
+                code_init += 'cuda_array_fill_{0}<<<1,1>>>({1}, {2});\n'.format(dtype, self._print(rhs.fill_value), self._print(lhs))
+        return code_init
+
+    def copy_CudaArray_Data(self, expr):
+        """ print the assignment of a Cuda NdArray
+
+        parameters
+        ----------
+            expr : PyccelAstNode
+                The Assign Node used to get the lhs and rhs
+        Return
+        ------
+            String
+                Return a str that contains the declaration of a dummy data_buffer
+                       and a call to an operator which copies it to a Cuda NdArray struct
+                if the ndarray is a stack_array the str will contain the initialization
+        """
+        rhs = expr.rhs
+        lhs = expr.lhs
+        if rhs.rank == 0:
+            raise NotImplementedError(str(expr))
+        dummy_array_name = self.scope.get_new_name('cuda_array_dummy')
+        declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+        dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
+        arg = rhs.arg if isinstance(rhs, (CudaArray, CupyArray)) else rhs
+        if rhs.rank > 1:
+            # flattening the args to use them in C initialization.
+            arg = self._flatten_list(arg)
+
+        self.add_import(c_imports['string'])
+        if isinstance(arg, Variable):
+            arg = self._print(arg)
+            cpy_data = "cudaMemcpy({0}.raw_data, {1}.{2}, {0}.buffer_size, cudaMemcpyHostToDevice);".format(lhs, arg, dtype)
+            return '%s\n' % (cpy_data)
+        else :
+            arg = ', '.join(self._print(i) for i in arg)
+            dummy_array = "%s %s[] = {%s};\n" % (declare_dtype, dummy_array_name, arg)
+            cpy_data = "cudaMemcpy({0}.raw_data, {1}, {0}.buffer_size, cudaMemcpyHostToDevice);".format(self._print(lhs), dummy_array_name, dtype)
+            return  '%s%s\n' % (dummy_array, cpy_data)
+
+    def _print_CudaSynchronize(self, expr):
+        return 'cudaDeviceSynchronize()'
+
+    def _print_CudaInternalVar(self, expr):
+        var_name = type(expr).__name__
+        var_name = cuda_Internal_Var[var_name]
+        dim_c = ('x', 'y', 'z')[expr.dim]
+        return '{}.{}'.format(var_name, dim_c)
+    
+    def cudaCopy(self, lhs, rhs):
+        from_location = 'Host'
+        to_location = 'Host'
+        if rhs.arg.memory_location in ('device', 'managed'):
+            from_location = 'Device'
+        if rhs.memory_location in ('device', 'managed'):
+            to_location = 'Device'
+        transfer_type = 'cudaMemcpy{0}To{1}'.format(from_location, to_location)
+        if isinstance(rhs.is_async, LiteralTrue):
+            cpy_data = "cudaMemcpyAsync({0}.raw_data, {1}.raw_data, {0}.buffer_size, {2}, 0);".format(lhs, rhs.arg, transfer_type)
```

Comment:
See above https://github.com/pyccel/pyccel/blob/master/developer_docs/codegen_stage.md#use-of-c-pointers